--- wine-11.14/dlls/ole32/compobj.c
+++ reactos/dll/win32/ole32/compobj.c
@@ -54,7 +54,9 @@
 #include "dde.h"
 #include "servprov.h"

+#ifndef __REACTOS__
 #include "initguid.h"
+#endif
 #include "compobj_private.h"
 #include "moniker.h"

@@ -253,7 +255,11 @@ static HKEY create_classes_root_hkey(DWORD access)
     attr.Length = sizeof(attr);
     attr.RootDirectory = 0;
     attr.ObjectName = &name;
+#ifdef __REACTOS__
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+#else
     attr.Attributes = 0;
+#endif
     attr.SecurityDescriptor = NULL;
     attr.SecurityQualityOfService = NULL;
     if (create_key( &hkey, access, &attr )) return 0;
@@ -302,7 +308,11 @@ LSTATUS create_classes_key( HKEY hkey, const WCHAR *na
     attr.Length = sizeof(attr);
     attr.RootDirectory = hkey;
     attr.ObjectName = &nameW;
+#ifdef __REACTOS__
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+#else
     attr.Attributes = 0;
+#endif
     attr.SecurityDescriptor = NULL;
     attr.SecurityQualityOfService = NULL;
     RtlInitUnicodeString( &nameW, name );
@@ -320,7 +330,11 @@ LSTATUS open_classes_key( HKEY hkey, const WCHAR *name
     attr.Length = sizeof(attr);
     attr.RootDirectory = hkey;
     attr.ObjectName = &nameW;
+#ifdef __REACTOS__
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+#else
     attr.Attributes = 0;
+#endif
     attr.SecurityDescriptor = NULL;
     attr.SecurityQualityOfService = NULL;
     RtlInitUnicodeString( &nameW, name );
--- wine-11.14/dlls/ole32/ole2.c
+++ reactos/dll/win32/ole32/ole2.c
@@ -70,6 +70,9 @@ typedef struct tagTrackerWindowInfo
   BOOL       escPressed;
   HWND       curTargetHWND;	/* window the mouse is hovering over */
   IDropTarget* curDragTarget;
+#ifdef __REACTOS__
+  HWND       accepterHWND;
+#endif
   POINTL     curMousePos;       /* current position of the mouse in screen coordinates */
   DWORD      dwKeyState;        /* current state of the shift and ctrl keys and the mouse buttons */
 } TrackerWindowInfo;
@@ -282,6 +285,13 @@ static inline BOOL is_droptarget(HWND hwnd)
     return get_droptarget_handle(hwnd) != 0;
 }

+#ifdef __REACTOS__
+static inline BOOL is_acceptfiles(HWND hwnd)
+{
+    return !!(GetWindowLong(hwnd, GWL_EXSTYLE) & WS_EX_ACCEPTFILES);
+}
+#endif
+
 /*************************************************************
  *           get_droptarget_local_handle
  *
@@ -762,6 +772,9 @@ HRESULT WINAPI DoDragDrop (
   trackerInfo.escPressed        = FALSE;
   trackerInfo.curTargetHWND     = 0;
   trackerInfo.curDragTarget     = 0;
+#ifdef __REACTOS__
+  trackerInfo.accepterHWND      = NULL;
+#endif

   hwndTrackWindow = CreateWindowW(OLEDD_DRAGTRACKERCLASS, L"TrackerWindow",
                                   WS_POPUP, CW_USEDEFAULT, CW_USEDEFAULT,
@@ -2177,14 +2190,107 @@ static LRESULT WINAPI OLEDD_DragTrackerWindowProc(
   return DefWindowProcW (hwnd, uMsg, wParam, lParam);
 }

+#ifdef __REACTOS__
+static HRESULT WINAPI DefaultDragEnter(HWND hwndTarget,
+                                       IDataObject* pDataObj,
+                                       DWORD grfKeyState,
+                                       POINTL pt,
+                                       DWORD* pdwEffect)
+{
+    HRESULT hr;
+    FORMATETC fme;
+
+    ZeroMemory(&fme, sizeof(fme));
+    fme.cfFormat = CF_HDROP;
+    fme.ptd = NULL;
+    fme.dwAspect = DVASPECT_CONTENT;
+    fme.lindex = -1;
+    fme.tymed = TYMED_HGLOBAL;
+    hr = pDataObj->lpVtbl->QueryGetData(pDataObj, &fme);
+
+    *pdwEffect = SUCCEEDED(hr) ? DROPEFFECT_COPY : DROPEFFECT_NONE;
+
+    if (*pdwEffect == DROPEFFECT_NONE)
+        return DRAGDROP_S_CANCEL;
+
+    return S_OK;
+}
+
+static HRESULT WINAPI DefaultDrop(HWND hwndAccepter,
+                                  IDataObject* pDataObj,
+                                  DWORD grfKeyState,
+                                  POINTL pt,
+                                  DWORD* pdwEffect)
+{
+    FORMATETC fme;
+    STGMEDIUM stgm;
+    HRESULT hr;
+    HGLOBAL hGlobal = NULL;
+
+    ZeroMemory(&fme, sizeof(fme));
+    fme.cfFormat = CF_HDROP;
+    fme.ptd = NULL;
+    fme.dwAspect = DVASPECT_CONTENT;
+    fme.lindex = -1;
+    fme.tymed = TYMED_HGLOBAL;
+    hr = pDataObj->lpVtbl->QueryGetData(pDataObj, &fme);
+    if (FAILED(hr))
+        return hr;
+
+    ZeroMemory(&stgm, sizeof(stgm));
+    hr = pDataObj->lpVtbl->GetData(pDataObj, &fme, &stgm);
+    if (SUCCEEDED(hr))
+    {
+        hGlobal = stgm.hGlobal;
+        if (hGlobal)
+        {
+            if (IsWindowUnicode(hwndAccepter))
+                PostMessageW(hwndAccepter, WM_DROPFILES, (WPARAM)hGlobal, 0);
+            else
+                PostMessageA(hwndAccepter, WM_DROPFILES, (WPARAM)hGlobal, 0);
+        }
+        ReleaseStgMedium(&stgm);
+    }
+
+    return hr;
+}
+#endif
+
 static void drag_enter( TrackerWindowInfo *info, HWND new_target )
 {
     HRESULT hr;
+#ifdef __REACTOS__
+    DWORD dwEffect = *info->pdwEffect;
+#endif

     info->curTargetHWND = new_target;

+#ifdef __REACTOS__
+    info->accepterHWND = NULL;
     while (new_target && !is_droptarget( new_target ))
+    {
+        if (is_acceptfiles(new_target))
+        {
+            dwEffect = info->dwOKEffect;
+            hr = DefaultDragEnter(new_target, info->dataObject,
+                                  info->dwKeyState, info->curMousePos,
+                                  &dwEffect);
+            dwEffect &= info->dwOKEffect;
+
+            if (hr == S_OK)
+            {
+                info->accepterHWND = new_target;
+                info->curDragTarget = NULL;
+                *info->pdwEffect = dwEffect;
+                return;
+            }
+        }
         new_target = GetParent( new_target );
+    }
+#else
+    while (new_target && !is_droptarget( new_target ))
+        new_target = GetParent( new_target );
+#endif

     info->curDragTarget = get_droptarget_pointer( new_target );

@@ -2202,6 +2308,9 @@ static void drag_enter( TrackerWindowInfo *info, HWND
             IDropTarget_Release( info->curDragTarget );
             info->curDragTarget = NULL;
             info->curTargetHWND = NULL;
+#ifdef __REACTOS__
+            info->accepterHWND = NULL;
+#endif
         }
     }
 }
@@ -2234,6 +2343,27 @@ static void drag_end( TrackerWindowInfo *info )
         IDropTarget_Release( info->curDragTarget );
         info->curDragTarget = NULL;
     }
+#ifdef __REACTOS__
+    else if (info->accepterHWND)
+    {
+        if (info->returnValue == DRAGDROP_S_DROP &&
+            *info->pdwEffect != DROPEFFECT_NONE)
+        {
+            *info->pdwEffect = info->dwOKEffect;
+            hr = DefaultDrop(info->accepterHWND, info->dataObject, info->dwKeyState,
+                             info->curMousePos, info->pdwEffect);
+            *info->pdwEffect &= info->dwOKEffect;
+
+            if (FAILED( hr ))
+                info->returnValue = hr;
+        }
+        else
+        {
+            *info->pdwEffect = DROPEFFECT_NONE;
+        }
+        info->accepterHWND = NULL;
+    }
+#endif
     else
         *info->pdwEffect = DROPEFFECT_NONE;
 }
@@ -2244,8 +2374,13 @@ static HRESULT give_feedback( TrackerWindowInfo *info
     int res;
     HCURSOR cur;

+#ifdef __REACTOS__
+    if (info->curDragTarget == NULL && info->accepterHWND == NULL)
+        *info->pdwEffect = DROPEFFECT_NONE;
+#else
     if (info->curDragTarget == NULL)
         *info->pdwEffect = DROPEFFECT_NONE;
+#endif

     hr = IDropSource_GiveFeedback( info->dropSource, *info->pdwEffect );

@@ -2311,6 +2446,9 @@ static void OLEDD_TrackStateChange(TrackerWindowInfo*
       trackerInfo->curDragTarget = NULL;
       trackerInfo->curTargetHWND = NULL;
     }
+#ifdef __REACTOS__
+    trackerInfo->accepterHWND = NULL;
+#endif

     if (hwndNewTarget)
       drag_enter( trackerInfo, hwndNewTarget );
@@ -2330,6 +2468,12 @@ static void OLEDD_TrackStateChange(TrackerWindowInfo*
                            trackerInfo->pdwEffect);
       *trackerInfo->pdwEffect &= trackerInfo->dwOKEffect;
     }
+#ifdef __REACTOS__
+    else if (trackerInfo->accepterHWND)
+    {
+      *trackerInfo->pdwEffect = trackerInfo->dwOKEffect;
+    }
+#endif
     give_feedback( trackerInfo );
   }
   else
The four cursor files remain at the older ReactOS state to avoid CORE-14608.
